hp_list/main.c

```diff
@@ -14,6 +14,92 @@
 #include <string.h>
 #include <threads.h>
 
+#ifdef RUNTIME_STAT
+/*
+ * "rtry" the number of retries in the __list_find function.
+ * "cons" the number of wait-free contains in the __list_find function that curr
+ * pointer pointed.
+ * "trav" the number of list element traversal in the __list_find function.
+ * "fail" the number of CAS() failures.
+ * "del" the number of list_delete operation failed and restart again.
+ * "ins" the number of list_insert operation failed and restart again.
+ * "load"  is the number of atomic_load operation in list_delete, list_insert
+ * and __list_find.
+ * "store" is the number of atomic_store operation in list_delete, list_insert
+ * and __list_find.
+ */
+static atomic_uint_fast64_t rtry = 0, cons = 0, trav = 0, fail = 0;
+static atomic_uint_fast64_t del = 0, ins = 0;
+static atomic_uint_fast64_t load = 0, store = 0;
+static atomic_uint_fast64_t deletes = 0, creates = 0;
+
+#define CAS(obj, expected, desired)                                          \
+    ({                                                                       \
+        bool __ret = atomic_compare_exchange_strong(obj, expected, desired); \
+        if (!__ret)                                                          \
+            atomic_fetch_add(&fail, 1);                                      \
+        __ret;                                                               \
+    })
+#define ATOMIC_LOAD(obj)            \
+    ({                              \
+        atomic_fetch_add(&load, 1); \
+        atomic_load(obj);           \
+    })
+#define ATOMIC_STORE_EXPLICIT(obj, desired, order)  \
+    do {                                            \
+        atomic_fetch_add(&store, 1);                \
+        atomic_store_explicit(obj, desired, order); \
+    } while (0)
+#define ATOMIC_FETCH_ADD(obj, arg, ops) ({ atomic_fetch_add(obj, 1); })
+
+void do_analysis(void)
```

Comment:
Instead of calling `do_analysis` at the end of `main` function explicitly, you can make use of [atexit(3)](https://man7.org/linux/man-pages/man3/atexit.3.html) to register `do_analysis` function. That is, we can call `register_stat()` at the beginning of `main`, and then `do_analysis` will be invoked automatically when the process is about to be terminated.

Reply:
Because the `atomic_fetch_add` will be only used with recording the operation state in related function in linked list,
 I think `ops` just mark  the meaning of  `ATOMIC_FETCH_ADD` is doing.